llvm/lib/Target/SystemZ/SystemZISelLowering.cpp

```diff
@@ -8030,7 +8077,431 @@ SDValue SystemZTargetLowering::combineBSWAP(
+        AddConst = AddConst & 0x3;
+      } else
+        AddConst = ~AddConst + 1;
+      // As original CCMask of of SELECT_CCMASK/BR_CCMASK does not have
+      // <= or >=.
+      CCVal &= 0x3;
+      CCVal += AddConst;
+      if (convertCCValToCCMask(CCVal)) {
+        // CCVal can not zero here.
+        CCMask ^= SystemZ::CCMASK_CMP_EQ;
+        if (Invert)
+          CCMask ^= SystemZ::CCMASK_ANY;
+        return true;
+      }
+    }
+    CCValid = RestoreCCValid;
+    return false;
+  }
+
+  // Optimize (ICMP (XOR (OP1 OP2))), OP1 or OP2 could be XOR again.
+  // One or both of operands could be (SELECT_CCMASK (ICMP (SRL (IPM (CC))))).
+  if (LHS->getOpcode() == ISD::XOR) {
+    SDValue XORReg = CCReg->getOperand(0);
+    bool Invert = false;
+    if (CCMask == SystemZ::CCMASK_CMP_NE)
+      Invert = !Invert;
+    // If both the operands are select_cc.
+    if (combineSRL_IPM_CCMask(XORReg, CCValid, CCMask)) {
+      CCReg = XORReg;
+      CCValid = SystemZ::CCMASK_ANY;
+      return true;
+    }
+    // Handle the case when one of the operand is select_cc and other operand
+    // could be xor again having both operands as select_cc.
+    auto *XOROp1 = LHS->getOperand(0).getNode();
+    auto *XOROp2 = LHS->getOperand(1).getNode();
+    if (!XOROp1 || !XOROp2)
+      return false;
+    if (XOROp1->getOpcode() == SystemZISD::SELECT_CCMASK ||
+        XOROp2->getOpcode() == SystemZISD::SELECT_CCMASK) {
+      auto *XOROp =
+          XOROp1->getOpcode() == SystemZISD::SELECT_CCMASK ? XOROp1 : XOROp2;
+      auto *CCMaskNode = dyn_cast<ConstantSDNode>(XOROp->getOperand(3));
+      auto *CCValidNode = dyn_cast<ConstantSDNode>(XOROp->getOperand(2));
+      if (!CCValidNode || !CCMaskNode)
+        return false;
+      int CCValidVal = CCValidNode->getZExtValue();
+      int CCMaskVal = CCMaskNode->getZExtValue();
+      SDValue XORReg1 = XOROp->getOperand(4);
+      SDValue XORReg2 = LHS->getOperand(1);
+      int CCMaskVal1 = CCMaskVal, CCMaskVal2 = CCMaskVal;
+      if (combineSRL_IPM_CCMask(XORReg1, CCValidVal, CCMaskVal1) &&
+          combineSRL_IPM_CCMask(XORReg2, CCValidVal, CCMaskVal2)) {
+        CCMask = CCMaskVal1 ^ CCMaskVal2;
+        CCReg = XORReg1;
+        CCValid = SystemZ::CCMASK_ANY;
+        return true;
+      }
+    }
+  }
+  CCValid = RestoreCCValid;
+  return false;
+}
+
 static bool combineCCMask(SDValue &CCReg, int &CCValid, int &CCMask) {
+  // combineSRL_IPM_CCMask tries to combine srl/ipm/cc sequence.
+  // This sequence here seems to be only for flag output operand.
+  // IPM operand has physical operand SystemZ::CC and CCValid is 15.
+  if (combineSRL_IPM_CCMask(CCReg, CCValid, CCMask))
+    return true;
+
+  // Code for SELECT_CCMASK does not seem to have ipm sequence.
+  // There is one case with sra/ipm that does not have SystemZ::CC as an
+  // operand. Test cases for sra/ipm are bcmp.ll, memcmp-01.ll and
+  // strcmp-01.ll. These tests have sra/sll/ipm/clc sequence.
```

Comment:
Well, those result from `addIPMSequence` in SystemZSelectionDAGInfo.cpp, which transforms the CC value from a comparison into a zero/positive/negative result as expected e.g. from `memcmp`.   Those will indeed have to remain a separate optimization.

However, it might still be preferable to integrate the contents of `combineSRL_IPM_CCMask` into this function at appropriate places.